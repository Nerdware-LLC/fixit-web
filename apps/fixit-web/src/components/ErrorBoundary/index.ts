export * from "./ErrorBoundary.jsx";

export * from "./classNames.js";
