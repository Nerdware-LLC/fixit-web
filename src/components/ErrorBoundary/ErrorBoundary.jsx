import React from "react";
import * as Sentry from "@sentry/react";
import { any, string } from "../../types";
import { logger } from "../../utils";

// NOTE: ErrorBoundary WON'T catch event-driven errors NOR async errors. So this will not work on, say, fetch state.

export class ErrorBoundary extends React.Component {
  constructor(props) {
    super(props);
    this.state = { error: null };
  }

  static getDerivedStateFromError(error) {
    // Update state so the next render will show the fallback UI.
    return { hasError: true };
  }

  componentDidCatch(error, errorInfo) {
    // You can also log the error to an error reporting service
    logger.error(
      error,
      `ErrorBoundary${
        this.props.identifier ? `:${this.props.identifier}` : null
      }`
    );
  }

  render() {
    return (
      <Sentry.ErrorBoundary showDialog {...this.props}>
        {this.props.children}
      </Sentry.ErrorBoundary>
    );
  }
}

ErrorBoundary.propTypes = {
  identifier: string,
  children: any.isRequired
};
