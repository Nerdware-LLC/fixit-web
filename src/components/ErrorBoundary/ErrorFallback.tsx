import Text from "@mui/material/Typography";

export const ErrorFallback = ({
  errorMessage = `Whoops! Something went wrong  :(\nPlease try again later.`,
}: ErrorFallbackProps) => (
  <div
    style={{
      height: "100%",
      display: "flex",
      alignItems: "center",
      justifyContent: "center",
      textAlign: "center",
      whiteSpace: "pre-wrap",
    }}
  >
    <Text>{errorMessage}</Text>
  </div>
);

export type ErrorFallbackProps = { errorMessage?: string };
