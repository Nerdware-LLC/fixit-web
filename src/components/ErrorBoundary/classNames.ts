/**
 * Class names for `ErrorBoundary` components (src/components/ErrorBoundary/).
 */
export const errorBoundaryClassNames = {
  defaultErrorFallbackRoot: "error-boundary__default-error-fallback-root",
} as const;
