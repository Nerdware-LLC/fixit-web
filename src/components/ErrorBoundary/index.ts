export * from "./ErrorBoundary";
export * from "./classNames";
