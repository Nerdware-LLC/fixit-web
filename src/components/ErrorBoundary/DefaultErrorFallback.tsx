import Box, { type BoxProps } from "@mui/material/Box";
import Text from "@mui/material/Typography";
import { errorBoundaryClassNames } from "./classNames";
import type { Except } from "type-fest";

/**
 * A default `fallback` component for `ErrorBoundary`.
 */
export const DefaultErrorFallback = ({
  errorMessage,
  style = {},
  ...boxProps
}: DefaultErrorFallbackProps) => (
  <Box
    className={errorBoundaryClassNames.defaultErrorFallbackRoot}
    style={{
      height: "100%",
      display: "flex",
      alignItems: "center",
      justifyContent: "center",
      textAlign: "center",
      whiteSpace: "pre-wrap",
      ...style,
    }}
    {...boxProps}
  >
    <Text>{errorMessage || `Whoops! Something went wrong  :(\nPlease try again later.`}</Text>
  </Box>
);

export type DefaultErrorFallbackProps = {
  errorMessage?: string;
} & Except<BoxProps, "children" | "className">;
