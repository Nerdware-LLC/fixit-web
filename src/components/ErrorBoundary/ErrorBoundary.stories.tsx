import { useState } from "react";
import Box from "@mui/material/Box";
import Button from "@mui/material/Button";
import { getMuiPaperStyles } from "@/app/ThemeProvider/helpers";
import { Dialog } from "@/components/Dialog";
import { DefaultErrorFallback } from "./DefaultErrorFallback";
import { ErrorBoundary } from "./ErrorBoundary";
import type { Meta, StoryObj } from "@storybook/react";

const meta = {
  title: "Components/ErrorBoundary",
  component: ErrorBoundary,
  tags: ["autodocs"],
  /* To use the ErrorBoundary component in a Story, we need a component that throws an
  error instead of returning JSX. This is accomplished in the below `render` fn using a
  state param called `shouldThrow` - this makes it necessary to implement ErrorBoundary
  as a `decorator` to ensure it is always rendered as a parent of the "Throw Error" btn. */
  decorators: [
    (Story, { args }) => (
      <Box
        component="fieldset"
        maxWidth="md"
        sx={{
          width: "21rem",
          height: "12rem",
          border: "2px dashed red",
          borderRadius: "0.5rem",
          padding: 0,
          display: "flex",
          alignItems: "center",
          justifyContent: "center",
          "& > legend": {
            whiteSpace: "nowrap",
            padding: "0 7px",
            margin: "auto",
          },
        }}
      >
        <legend>ErrorBoundary visual representation</legend>
        <ErrorBoundary {...args}>
          <Story />
        </ErrorBoundary>
      </Box>
    ),
  ],
  render: () => {
    const [shouldThrow, setShouldThrow] = useState(false);
    if (shouldThrow) throw new Error();
    return <Button onClick={() => setShouldThrow(true)}>Throw an Error</Button>;
  },
} satisfies Meta<typeof ErrorBoundary>;

export default meta;

///////////////////////////////////////////////////////////
// STORIES

type Story = StoryObj<typeof meta>;

export const BasicDemo = {
  args: {
    fallback: ({ resetError }) => (
      <Dialog
        title={`You clicked the "Throw Error" button!`}
        message={
          <Box
            component="span"
            sx={(theme) => ({
              "& > code": {
                padding: "2px 4px",
                borderRadius: "3px",
                ...getMuiPaperStyles(1, theme),
              },
            })}
          >
            More info regarding the error and its handling is available in the <code>console</code>,
            as well as the Storybook <code>Actions</code> tab.
          </Box>
        }
        acceptLabel="OK - Reset Error"
        handleAccept={() => resetError()}
      />
    ),
  },
} satisfies Story;

export const WithDefaultFallback = {
  args: {
    fallback: ({ resetError }) => (
      <DefaultErrorFallback
        onClick={() => resetError()}
        style={{ width: "100%", height: "100%" }}
      />
    ),
  },
} satisfies Story;
