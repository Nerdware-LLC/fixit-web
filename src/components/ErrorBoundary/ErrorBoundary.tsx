import { getErrorMessage } from "@nerdware/ts-type-safety-utils";
import * as Sentry from "@sentry/react";
import { logger } from "@/utils/logger.js";
import { DefaultErrorFallback } from "./DefaultErrorFallback.jsx";

/**
 * Default `onError` handler fn for {@link ErrorBoundary}
 */
const defaultErrorBoundaryOnErrorHandler: ErrorBoundaryOnErrorHandler = (
  error: Error,
  _componentStack: string,
  _eventId: string
) => {
  logger.error(error, "Sentry.ErrorBoundary");
};

/**
 * Default `fallback` render fn for {@link ErrorBoundary}
 */
const defaultErrorBoundaryFallback: ErrorBoundaryFallbackFn = ({ error }) => (
  <DefaultErrorFallback errorMessage={getErrorMessage(error)} />
);

/**
 ;* A reusable [`Sentry ErrorBoundary`][sentry-eb-docs] component.
 *
 * > **NOTE:** ErrorBoundary components DO NOT catch the following:
 * >   - Event-driven errors
 * >   - Uncaught promise rejections
 *
 * #### [Sentry.ErrorBoundary Options][sentry-eb-opts]:
 *
 * - `showDialog` - If a [Sentry User Feedback Widget][sentry-feedback-widget] should be rendered
 *   when the ErrorBoundary catches an error.
 *
 *   - _**Default:**_ `true`
 *
 * - `dialogOptions` - Options that are passed into the Sentry User Feedback Widget. See all
 *   possible customization options [here][sentry-feedback-widget-opts].
 *
 * - `fallback` - JSX to render when the ErrorBoundary catches an error. Can be JSX or a fn that
 *   returns JSX. If you provide a fn, Sentry will call it with additional info and helpers.
 *
 *   - _**Default:**_ {@link DefaultErrorFallback|`DefaultErrorFallback`}
 *
 * - `onError` - A fn that gets called when the ErrorBoundary encounters an error. `onError` is
 *   useful if you want to propagate the error into a state management library like Redux, or if
 *   you want to check any side effects that could have occurred due to the error.
 *
 *   - _**Default:**_ {@link logger.error|`logger.error`}
 *
 * - `onMount` - A fn that gets called in `componentDidMount()`.
 *
 * - `onUnmount` - A fn that gets called in `componentWillUnmount()`.
 *
 * - `beforeCapture` - A fn that gets called before an error is sent to Sentry, allowing for extra
 *   tags or context to be added to the error.
 *
 * Click [here to view source code][sentry-eb-src] for `Sentry.ErrorBoundary`.
 *
 * @docs
 * - [React docs: Error Boundaries][react-eb-docs]
 * - [Sentry.ErrorBoundary][sentry-eb-docs]
 * - [Sentry.ErrorBoundary Options][sentry-eb-opts]
 *
 * [react-eb-docs]: https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary
 * [sentry-eb-src]: https://github.com/getsentry/sentry-javascript/blob/develop/packages/react/src/errorboundary.tsx
 * [sentry-eb-docs]: https://docs.sentry.io/platforms/javascript/guides/react/features/error-boundary/
 * [sentry-eb-opts]: https://docs.sentry.io/platforms/javascript/guides/react/features/error-boundary/#options
 * [sentry-feedback-widget]: https://docs.sentry.io/platforms/javascript/guides/react/enriching-events/user-feedback/
 * [sentry-feedback-widget-opts]: https://docs.sentry.io/platforms/javascript/guides/react/enriching-events/user-feedback/#customizing-the-widget
 */
export class ErrorBoundary extends Sentry.ErrorBoundary {
  public static readonly defaultProps = {
    showDialog: true,
    onError: defaultErrorBoundaryOnErrorHandler,
    fallback: defaultErrorBoundaryFallback,
  } satisfies Partial<Sentry.ErrorBoundaryProps>;
}

/**
 * `onError` handler fn for {@link ErrorBoundary}
 * @note Available `onError` fn args: `error`, `componentStack`, `eventId`
 */
export type ErrorBoundaryOnErrorHandler = NonNullable<Sentry.ErrorBoundaryProps["onError"]>;

/**
 * `fallback` render fn for {@link ErrorBoundary}
 * @note Available `fallback` fn args: { `error`, `componentStack`, `eventId`, `resetError` }
 */
export type ErrorBoundaryFallbackFn = NonNullable<Sentry.ErrorBoundaryProps["fallback"]>;
