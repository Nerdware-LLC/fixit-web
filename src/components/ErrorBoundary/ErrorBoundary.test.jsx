import renderer from "react-test-renderer";
import { ErrorBoundary } from "./ErrorBoundary";

it("renders correctly", () => {
  const tree = renderer
    .create(
      <ErrorBoundary>
        <div>
          <p>This is a TEST of ErrorBoundary.</p>
        </div>
      </ErrorBoundary>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
