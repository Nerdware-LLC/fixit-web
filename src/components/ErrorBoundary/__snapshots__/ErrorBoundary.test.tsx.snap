// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders correctly 1`] = `
<div>
  <p>
    This is a TEST of ErrorBoundary.
  </p>
</div>
`;
